---
title: "Spatial Transcriptomics Part 1: Data Exploration"
author: "UCD Bioinformatics Core"
date: "`r Sys.Date()`"
output:
    html_document:
      keep_md: TRUE
      toc: TRUE
---

# Spatial Transcriptomics Part 1: Data Exploration

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, error = FALSE)
```

## Read the Xenium output into Seurat


```{r libraries}
library(Seurat)     # Spatial transcriptomics analysis
library(kableExtra) # format tables
library(ggplot2)   # create graphics
library(viridis)   # accessible color palettes
```

## Create Seurat object
[Seurat](http://satijalab.org/seurat/) is a popular R package that is designed for QC, analysis, and exploration of single cell data, and spatial single cell data. We will use Seurat in the first stage in analysis to carry out a series of standard procedures: read in spatial data, QC, normalization, dimensionality reduction, clustering and visualizations.
Seurat has an extensive documentation that covers many different use cases. In addition to the standard Seurat workflow, this documentation makes use of some custom code, and brings in functions from other packages. For additional information on Seurat standard workflows, see the authors' [tutorials](https://satijalab.org/seurat/articles/get_started_v5_new).

#### Read in spatial data and cell feature matrix

First, we read in data from each individual tissue folder and create Seurat objects.

```{r load_data}
## set up directories
data.dir <- "./00-RawData"
project.dir <- "./"

## samples
samples <- c("TgCRND8", "wildtype")
sample.dir <- paste0("Xenium_V1_FFPE_", samples, "_5_7_months_outs")

## modify "future.globals.maxSize" to accommodate larege data"
options(future.globals.maxSize = 2000 * 1024 ^ 2)
## load multiple slices
data.xenium <- lapply(seq_along(samples), function(i){
	## modify the fov slot names to distinguish fovs from different tissue samples for easier access later
	tmp <- Seurat::LoadXenium(data.dir = file.path(data.dir, sample.dir[i]), segmentations = "cell", cell.centroids = T, fov = paste0("fov.", samples[i]))
	## attach tissue sample name to cell names
	tmp <- RenameCells(tmp, new.names = paste0(samples[i], "-", Cells(tmp)))
	## add metadata to identify tissue samples
	tmp <- AddMetaData(tmp, metadata = samples[i], col.name = "Genotype")
	return(tmp)
})

for (i in seq_along(data.xenium)){
	if (i == 1){
		experiment.aggregate <- data.xenium[[i]]
	}else{
		experiment.aggregate <- merge(experiment.aggregate, data.xenium[[i]])
	}
}
```

## Explore the Seurat object

A Seurat object is a complex data structure containing the data from a spatial single cell assay and **all** of the information associated with the experiment, including annotations, analysis, and more. This data structure was developed by the authors of the Seurat analysis package, for use with their pipeline.

```{r view, eval=FALSE}
View(experiment.aggregate)
```

Most Seurat functions take the object as an argument, and return either a new Seurat object or a ggplot object (a visualization). As the analysis continues, more and more data will be added to the object.

```{r explore_object_1}
slotNames(experiment.aggregate)
experiment.aggregate@assays # a slot is accessed with the @ symbol
```

- Which slots are empty, and which contain data?
- What type of object is the content of the meta.data slot?
- What metadata is available?

There is often more than one way to interact with the information stored in each of a Seurat objects' many slots. The default behaviors of different access functions are described in the help documentation.

```{r explore_object_2}
# which slot is being accessed here? find another way to produce the result
head(experiment.aggregate[[]])
```

The use of syntax is often a matter of personal preference. In the interest of clarity, this documentation will generally use the more explicit syntax, with a few exceptions.

## QC visuzalization

```{r visual}
## genes: designed probes for the gene panel
VlnPlot(experiment.aggregate, features = c("nFeature_Xenium", "nCount_Xenium"), ncol = 2, pt.size = 0, log = T)

## blank, unassigned codewords: are unused codewords. No probe in the corresponding gene panel that will generate the codeword
VlnPlot(experiment.aggregate, features = c("nFeature_BlankCodeword", "nCount_BlankCodeword"), ncol = 2, pt.size = 0.01)

## negative control codewords: codewords that do not have any probes matching that code. They are used to assess the specificity of the decoding algoritm
VlnPlot(experiment.aggregate, features = c("nFeature_ControlCodeword", "nCount_ControlCodeword"), ncol = 2, pt.size = 0.01)

## control, negative control probe: probes that exist in the panel but do not target any biological sequences. They are used to assess the specificity of the assay
VlnPlot(experiment.aggregate, features = c("nFeature_ControlProbe", "nCount_ControlProbe"), ncol = 2, pt.size = 0.01)

## genomic control: designed to bind to intergenic genomic DNA but not to any transcript sequence present in the tissue. They are present in the Xenium Prime 5K assay, but not in other Xenium assays.
```

## Exploratory visualizations


```{r}
## plot expression for genes of interest
ImageDimPlot(experiment.aggregate, fov = "fov.wildtype", molecules = c("Satb2", "Lamp5", "Car4"), group.by = NULL, size = 0.5, alpha = 0.5, axes = T)
ImageDimPlot(experiment.aggregate, fov = "fov.TgCRND8", molecules = c("Satb2", "Lamp5", "Car4"), group.by = NULL, size = 0.5, alpha = 0.5, axes = T)
ImageFeaturePlot(experiment.aggregate, fov = "fov.wildtype", features = c("Satb2", "Lamp5", "Car4"), size = 0.75, cols = c("white", "red"))
ImageFeaturePlot(experiment.aggregate, fov = "fov.TgCRND8", features = c("Satb2", "Lamp5", "Car4"), size = 0.75, cols = c("white", "red"))

## zoom in on specific regions
wildtype.subfov.1 <- Crop(experiment.aggregate[["fov.wildtype"]], y = c(1300, 3500), x = c(1200, 4100), coords = "plot")
experiment.aggregate[["wildtype.subfov.1"]] <- wildtype.subfov.1
DefaultBoundary(experiment.aggregate[["wildtype.subfov.1"]]) <- "segmentation"
ImageDimPlot(experiment.aggregate, fov = "wildtype.subfov.1", axes = T, border.color = "white", border.size = 0.1, cols = "polychrome",
	coord.fixed = F, molecules = c("Satb2", "Lamp5", "Car4"), alpha = 0.5, nmols = 10000)
TgCRND8.subfov.1 <- Crop(experiment.aggregate[["fov.TgCRND8"]], y = c(3700, 4800), x = c(300, 1800), coords = "plot")
experiment.aggregate[["TgCRND8.subfov.1"]] <- TgCRND8.subfov.1
DefaultBoundary(experiment.aggregate[["TgCRND8.subfov.1"]]) <- "segmentation"
ImageDimPlot(experiment.aggregate, fov = "TgCRND8.subfov.1", axes = T, border.color = "white", border.size = 0.1, cols = "polychrome",
	coord.fixed = F, molecules = c("Satb2", "Lamp5", "Car4"), alpha = 0.5, nmols = 10000)

```

## Prepare for the next section

#### Save object
```{r saveRDS}
saveRDS(experiment.aggregate, file="Spatial_workshop-01.rds")
```

#### Download Rmd
```{r download_Rmd, eval=FALSE}
download.file("https://raw.githubusercontent.com/ucdavis-bioinformatics-training/2025-March-Spatial-Transcriptomics/main/data_analysis/02-Clustering.Rmd", "02-Clustering.Rmd")
```

#### Session information
```{r sessioinfo}
sessionInfo()
```
