---
title: "Spatial Transcriptomics Part 4: Simple Niche analysis"
author: "UCD Bioinformatics Core"
date: "`r Sys.Date()`"
output:
    html_document:
      keep_md: TRUE
      toc: TRUE
---

# Spatial Transcriptomics Part 4: Simply Niche analysis

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, error = FALSE)
```

## Set up the workspace


```{r libraries}
library(Seurat)     # Spatial transcriptomics analysis
library(kableExtra) # format tables
library(ggplot2)   # create graphics
library(viridis)   # accessible color palettes
library(ComplexHeatmap)  # visualization heatmaps
```

We are going to use the cortex region that was subsetted previously for this exercise.


```{r load_data}
cortex <- readRDS("cortex.rds")
```

## Find niches

Up until now, we have done analyses that are very similar to regular single cell RNASeq data analysis, without utilizing the spatial information of the cell except for in visualizations. One of the popular analyses for spatial transcriptomics is to identify niches. A nich are regions of tissue, each of which is defined by a different composition of spatially adjacent cell types. In Seurat, a local neighborhood for each cell is constructed by including its _neighbors.k_ spatially closest neighbors, and count the occurences of each cell type present in this neighborhood. Then a k-means clustering is done to group cells that have similar neighborhoods together, which is defined as spatial niches.

We will use _BuildNicheAssay_ function to perform this analysis.

```{r}
cortex <- BuildNicheAssay(cortex, fov = "fov.TgCRND8", group.by = "predicted.celltype", niches.k = 5, neighbors.k = 30)
```

Let's visualize the niche result together with the cell types.

```{r}
ImageDimPlot(cortex, fov = "fov.TgCRND8", group.by = "predicted.celltype", size = 0.75, dark.background = F)
ImageDimPlot(cortex, fov = "fov.TgCRND8", group.by = "niches", size = 0.75, dark.background = F)
```

Next, we can tally the number of each cell type within a niche.

```{r}
mat <- table(cortex$predicted.celltype, cortex$niches)
mat
```

Let's visualize this relationship

```{r}
cell_fun = function(j, i, x, y, width, height, fill) {
                grid::grid.rect(x = x, y = y, width = width *0.99, 
                                height = height *0.99,
                                gp = grid::gpar(col = "grey", 
                                                fill = fill, lty = 1, lwd = 0.5))
}

col_fun=circlize::colorRamp2(c(-2, 0, 2), c("blue", "white", "red"))

Heatmap(scale(mat), show_row_dend = F, show_column_dend = F, rect_gp = grid::gpar(type = "none"), cell_fun = cell_fun, col = col_fun, column_names_rot = 90)
```

Please play around with the number of niches that _BuildNicheAssay_ should find and see how the cell type distribution changes.

```{r}
library(SpatialExperiment)
library(SummarizedExperiment)
library(scuttle)
library(scater) 
library(cowplot) 
    

cortex <- readRDS("cortex.rds")
exp <- LayerData(cortex, layer = "counts")
coords <- GetTissueCoordinates(cortex, which = "centroids")
rownames(coords) <- coords$cell
coords <- as.matrix(coords[,1:2])
    
se <- SpatialExperiment(assay = list(counts = exp), spatialCoords = coords)

se <- computeLibraryFactors(se)

assay(se, "data") <- normalizeCounts(se)

k_geom <- c(15, 30)
se <- computeBanksy(se, compute_agf = T, k_geom = k_geom, assay_name = "data")

lambda <- c(0.2, 0.8)
set.seed(1000)
se <- runBanksyPCA(se, use_agf = T, lambda = lambda)
se <- runBanksyUMAP(se, use_agf = T, lambda = lambda)
se <- clusterBanksy(se, use_agf = T, lambda = lambda, resolution = c(0.3, 0.5, 0.8, 1))

cnames <- colnames(colData(se))
cnames <- cnames[grep("^clust", cnames)]
colData(se) <- cbind(colData(se), spatialCoords(se))

plot_lam02res08 <- plotColData(se,
    x = 'x', y = 'y',
    point_size = 0.2, colour_by = cnames[1]
)
plot_lam02res1 <- plotColData(se,
    x = "x", y = "y",
    point_size = 0.2, colour_by = cnames[2]
)
plot_lam08res08 <- plotColData(se,
    x = "x", y = "y",
    point_size = 0.2, colour_by = cnames[3]
)
plot_lam08res1 <- plotColData(se,
    x = "x", y = "y",
    point_size = 0.2, colour_by = cnames[4]
)
plot_lam02res03 <- plotColData(se,
    x = 'x', y = 'y',
    point_size = 0.2, colour_by = cnames[5]
)
plot_lam02res05 <- plotColData(se,
    x = "x", y = "y",
    point_size = 0.2, colour_by = cnames[6]
)
plot_lam08res03 <- plotColData(se,
    x = "x", y = "y",
    point_size = 0.2, colour_by = cnames[7]
)
plot_lam08res05 <- plotColData(se,
    x = "x", y = "y",
    point_size = 0.2, colour_by = cnames[8]
)



plot_grid(plotlist = list(plot_lam02res08 + coord_equal(), plot_lam02res1 + coord_equal(),
        plot_lam08res08 + coord_equal(), plot_lam08res1 + coord_equal(),
        plot_lam02res03 + coord_equal(), plot_lam02res05 + coord_equal(),
        plot_lam08res03 + coord_equal(), plot_lam08res05 + coord_equal()),
        ncol = 1)
dev.off()
```

## Prepare for the next section


#### Download Rmd
```{r download_Rmd, eval=FALSE}
download.file("https://raw.githubusercontent.com/ucdavis-bioinformatics-training/2025-March-Spatial-Transcriptomics/main/data_analysis/05-NicheDE.Rmd", "05-NicheDE.Rmd")
```

#### Session information
```{r sessioinfo}
sessionInfo()
```
